pkg/references/errors.go

```diff
@@ -0,0 +1,207 @@
+package references
+
+import (
+	"fmt"
+
+	"github.com/jexia/semaphore/pkg/lookup"
+	"github.com/jexia/semaphore/pkg/pretty_errors"
+	"github.com/jexia/semaphore/pkg/specs"
+)
+
+type wrapErr struct {
+	Inner error
+}
+
+func (i wrapErr) Unwrap() error {
+	return i.Inner
+}
+
+type (
+	// ErrUnresolvedFlow occurs when the whole flow cannot be resolved. It's the root error in this package.
+	ErrUnresolvedFlow struct {
+		wrapErr
+		Name string
+	}
+
+	// ErrUndefinedReference occurs when resolving meets unknown reference.
+	ErrUndefinedReference struct {
+		wrapErr
+		Expression specs.Expression
+		// Reference points to the reference which caused the error
+		Reference  *specs.PropertyReference
+		Breakpoint string
+		Path       string
+	}
+
+	// ErrUndefinedResource occurs when resolving meets unknown resource
+	ErrUndefinedResource struct {
+		wrapErr
+		// Reference points to the reference which caused the error
+		Reference  *specs.PropertyReference
+		Breakpoint string
+		// AvailableReferences contains the whole list of known references
+		AvailableReferences map[string]lookup.ReferenceMap
+	}
+
+	ErrUnresolvedOnError struct {
+		wrapErr
+		OnError *specs.OnError
+	}
+
+	ErrUnresolvedNode struct {
+		wrapErr
+		Node *specs.Node
+	}
+
+	ErrUnresolvedParameterMap struct {
+		wrapErr
+		Parameter *specs.ParameterMap
+	}
+
+	ErrUnresolvedProperty struct {
+		wrapErr
+		Property *specs.Property
+	}
+
+	ErrUnresolvedParams struct {
+		wrapErr
+		Params map[string]*specs.Property
+	}
+
+	ErrUnresolvedCall struct {
+		wrapErr
+		Call *specs.Call
+	}
+)
```

Comment:
Please keep a struct and its methods as close as possible to each other. I think it is best to move the methods underneath the structs to avoid having to search for implementations.

Reply:
well, I did as you suggest in the beginning, but I got in trouble soon: it was really hard to see the list of all the existing errors. This groupped `type` definition lets us to list all the errors together.

Reply:
Let's have a chat about how we are going to maintain and grow these errors. 😃